lib/sparse/src/index/search_context.rs

```diff
@@ -106,72 +126,49 @@ impl<'a> SearchContext<'a> {
         queue.into_vec()
     }
 
-    /// Advance posting lists iterators and return the next candidate by increasing ids.
-    ///
-    /// Example
-    ///
-    /// postings_iterators:
-    ///
-    /// 1,  30, 34, 60, 230
-    /// 10, 30, 35, 51, 230
-    /// 2,  21, 34, 60, 200
-    /// 2,  30, 34, 60, 230
-    ///
-    /// Next:
-    ///
-    /// a,  30, 34, 60, 230
-    /// 10, 30, 35, 51, 230
-    /// 2,  21, 34, 60, 200
-    /// 2,  30, 34, 60, 230
-    ///
-    /// Next:
-    ///
-    /// a,  30, 34, 60, 230
-    /// 10, 30, 35, 51, 230
-    /// b,  21, 34, 60, 200
-    /// b,  30, 34, 60, 230
-    ///
-    /// Next:
-    ///
-    /// a,  30, 34, 60, 230
-    /// c,  30, 35, 51, 230
-    /// b,  21, 34, 60, 200
-    /// b,  30, 34, 60, 230
-    fn advance(&mut self) -> Option<ScoredPointOffset> {
-        // Get current min record id from all posting list iterators
-        let current_min_record_id = self.min_record_id?;
-        let mut score = 0.0;
-        let mut found = false;
-        // Iterate to advance matching posting iterators
-        for posting_iterator in self.postings_iterators.iter_mut() {
-            if let Some(element) = posting_iterator.posting_list_iterator.peek() {
-                // accumulate score for the current record id
-                if element.record_id == current_min_record_id {
-                    found = true;
-                    score += element.weight * posting_iterator.query_weight;
-                    // advance posting list iterator to next element
-                    posting_iterator.posting_list_iterator.advance();
+    /// Advance posting lists iterators in a batch fashion.
+    fn advance_batch<F: Fn(PointOffsetType) -> bool>(
+        &mut self,
+        batch_start_id: PointOffsetType,
+        batch_last_id: PointOffsetType,
+        filter_condition: &F,
+    ) {
+        for posting in self.postings_iterators.iter_mut() {
+            for element in posting.posting_list_iterator.remaining_elements() {
+                let element_id = element.record_id;
+                if element_id > batch_last_id {
+                    // reaching end of the batch
+                    break;
                 }
-            } else {
-                // mark search context for cleanup if any posting list iterator is exhausted
-                self.contains_empty_posting = true;
+                let element_score = element.weight * posting.query_weight;
```

Comment:
I cannot understand why do we skip here the filter condition? What if `element.record_id` is filtered?

Reply:
Got it, we calculate score even for deleted points. I can propose to try avoid this additional calculation with doing filter conditions before scoring. keep filter result in biset and resuse it for both scoring and `self.result_queue` update

Reply:
after discussion with @agourlay we decided to keep as is because benchmark shown that this approach is better